Feature/add teamhub microservice

the-man-w-laughs opened the pull request:
Big-service which will be responsible for CRUD operations for projects, project_members, tasks, tasks_handlers, tasks_comments.
Updated docker-compose to run both services and dbs at once.
In future this service will or wont be divided into 2-3 separated services.

Tech-stack:
CQRS (MediatR);
Fluent validation;
Automapper.

the-man-w-laughs committed:
Add initial docker configuration.

the-man-w-laughs committed:
Add identity (and db) to docker-compose.

Major changes:
- change https to http;
- add cors config;
- add docker-compose.

the-man-w-laughs committed:
Add db models.

the-man-w-laughs committed:
Set CQRS pattern.

the-man-w-laughs committed:
Add repositories.

the-man-w-laughs committed:
Add endpoints templates.

the-man-w-laughs committed:
Add second service to docker-compose.

the-man-w-laughs committed:
Implement controllers.

the-man-w-laughs committed:
Refactor cqrs handlers.

katyalys commented on src/services/Identity/Identity.Application/ApplicationRegisterExtensions.cs:
```diff
@@ -0,0 +1,17 @@
+    {
+        public static void RegisterApplicationDependencies(this IServiceCollection services)
+        {
+            services.AddAutoMapper(typeof(UsersProfile));
```
Let it be for now, but in general when you will have many models and entities better to scan the executing assembly for AutoMapper profiles and not to write services.AddAutoMapper(typeof()) for every entity

katyalys commented on src/services/TeamHub/TeamHub.BLL/Extensions/ClaimsPrincipalExtensions.cs:
```diff
@@ -0,0 +1,34 @@
+    public static int GetUserId(this IHttpContextAccessor httpContextAccessor)
+    {
+        var principal = httpContextAccessor?.HttpContext?.User;
+        if (principal == null)
```
add empty line before if

katyalys commented on src/services/TeamHub/TeamHub.BLL/Extensions/ClaimsPrincipalExtensions.cs:
```diff
@@ -0,0 +1,34 @@
+        }
+
+        var claim = principal.FindFirst(x => x.Type.Equals(ClaimTypes.NameIdentifier));
+        if (claim == null)
```
add empty line before if

katyalys commented on src/services/TeamHub/TeamHub.BLL/MediatR/CQRS/Project/CommandHandlers/UpdateProjectCommandHandler.cs:
```diff
@@ -0,0 +1,48 @@
+    public async Task<int> Handle(UpdateProjectCommand request, CancellationToken cancellationToken)
+    {
+        var userId = _httpContextAccessor.GetUserId();
+        ;
```
delete comma

katyalys commented on src/services/TeamHub/TeamHub.DAL/Repositories/Repository.cs:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    public virtual async Task<List<TEntity>> GetAllAsync()
+    {
```
get all data from db is bad for perfomance, use pagination

katyalys commented on src/services/TeamHub/TeamHub.WebApi/controllers/ProjectsController.cs:
```diff
@@ -0,0 +1,165 @@
+    /// <summary>
+    /// Create New Team Member
+    /// </summary>
+    [HttpPost("{projectId:int}/TeamMembers/{userId:int}")]
```
use lowercase in route naming

katyalys commented on src/services/TeamHub/TeamHub.WebApi/controllers/ProjectsController.cs:
```diff
@@ -0,0 +1,165 @@
+    /// Delete TeamMember
+    /// </summary>
+    [HttpDelete("{projectId:int}/TeamMembers/{userId:int}")]
+    public async Task<IActionResult> DeleteTeamMember(
```
lowercase

katyalys commented on src/services/TeamHub/TeamHub.WebApi/controllers/ProjectsController.cs:
```diff
@@ -0,0 +1,165 @@
+    /// Create New Task
+    /// </summary>
+    [HttpPost("{projectId:int}/Tasks")]
+    public async Task<IActionResult> CreateNewTaskModel(
```
same

katyalys commented on src/services/TeamHub/TeamHub.WebApi/controllers/ProjectsController.cs:
```diff
@@ -0,0 +1,165 @@
+    /// Get All Project tasks
+    /// </summary>
+    [HttpGet("{projectId:int}/Tasks")]
+    public async Task<IActionResult> GetAllProjectsTaskModels([FromRoute] int projectId)
```
same

katyalys commented on src/services/TeamHub/TeamHub.WebApi/controllers/TaskModelController.cs:
```diff
@@ -0,0 +1,133 @@
+    /// Get All Task Handlers
+    /// </summary>
+    [HttpGet("{taskId:int}/TaskHandlers")]
+    public async Task<IActionResult> GetAllTaskHandlers([FromRoute] int taskId)
```
lowercase everywhere

katyalys commented on src/services/Identity/Identity.Infrastructure/Repositories/AppUserRepository.cs:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    public async Task<IEnumerable<AppUser>> GetAllUsersAsync()
+    {
```
use pagination

katyalys commented on src/services/Identity/Identity.WebAPI/Controllers/UsersController.cs:
```diff
@@ -0,0 +1,63 @@
+    [Authorize]
+    public async Task<IActionResult> GetAllUsersAsync()
+    {
+        var result = await _userService.GetAllUsersAsync();
```
add empty lines before each return in class

katyalys commented on src/services/TeamHub/TeamHub.BLL/AutoMapperProfiles/TaskModelProfile.cs:
```diff
@@ -0,0 +1,34 @@
+        private List<User> GetUsersList(ICollection<TaskHandler> teamMembers)
+        {
+            List<User> users = new List<User>();
+            foreach (var member in teamMembers)
```
empty line before foreach

the-man-w-laughs committed:
Move error handling from repositories to CQRS handlers.

katyalys commented on src/services/TeamHub/TeamHub.BLL/AutoMapperProfiles/ProjectsProfile.cs:
```diff
@@ -0,0 +1,33 @@
+        }
+
+        private List<User> GetUsersList(ICollection<TeamMember> teamMembers)
+        {
```
check if you really need this method, maybe you can use linq select in MapFrom

the-man-w-laughs committed:
Refactor code.

Major changes:
- Remove uppercase in routes;
- Add empty line before each 'if';
- Change functions to linq queries in Automapper profiles.

the-man-w-laughs committed:
Add pagination to Identity.

the-man-w-laughs committed:
Add pagination to TeamHub.

the-man-w-laughs committed:
Merge branch 'feature/add-identity' into feature/add-teamhub-microservice
